### DISCUSSION TITLE
Command not found

### DISCUSSION THREAD

**Discussion by Pumpapa** [Taitava/obsidian-shellcommands#367] [Q&A] [answered]
Just installed ShellCommands. It does work, but the example from the docs `gedit {{file_path:absolute}}` produces `/bin/bash: line 1: gedit: command not found`. 

When I try `/usr/bin/gedit {{file_path:absolute}}`, it produces `/bin/bash: line 1: /usr/bin/gedit: No such file or directory`. (This refers to the command /usr/bin/gedit; not the subject file).

I checked `echo $PATH`, which does include `/usr/bin`, where gedit is located.

It is weird. When I try `echo` or `ls -l` or even `less` (or /usr/bin/less) on the file, it works as it should. That is, in the `cntrl >_` it offers the expected result, and in Obsidian from the `>_` menu it doesn't give an error message and when I don't ignore `stdout` but make a balloon, it shows the expected result.

Any help is appreciated.

I'm working with Obsidian 1.3.7, Shell Commands 0.20.0, on Pop OS (Ubuntu clone) 22.04.

**Comment by Taitava** [comment]
Hmm, sounds weird.
1. Did you try to execute `/usr/bin/gedit` in a regular terminal? So not using this plugin. Just to see that gedit can be generally executed.
2. Are you able to execute some other program from `/usr/bin` using the SC plugin?
3. Can you provide some screenshots about the problem? Maybe there's some tiny detail I'm missing out that I could see from images.

**Reply by Pumpapa** [reply]
1. Yes, `gedit` works in a terminal
2. Yes. There's something odd though: I tried several console commands (`echo`, `ls -l`, `tail`, `wc`) and they all work in `cntrl >_`, but several commands that would open a new window (such as `gedit`, `drawio`, `thunderbirs`) don't work. They all give the error `command not found`
3. OK. I've uploaded one screenshot with the result of `ls -l`, which shows that `ls` can be executed and one with the output for `gedit`.

I poked around a bit more and I noticed a problem. 
When I execute `ls -l /usr/bin | wc` I get 717 lines, whereas when I execute it in a terminal, I get 2463 lines.
I compared the output of `ls /usr/bin ` in ShellCommands and in a terminal, but I don't understand the difference (3rd screenshot). Two directory entries for `less` and `gedit` are:
`-rwxr-xr-x 1 root root       199048 Feb  7  2023  less`
`-rwxr-xr-x 1 root root        14640 Mar  2  2022  gedit`
so the difference doesn't seem to be permissions or ownership.
I'm flabbergasted, and thanks for looking into this.

 
![Screenshot from 2023-08-24 20-10-45](https://github.com/Taitava/obsidian-shellcommands/assets/1771485/36fa3fc4-7081-4e4b-bbf2-8ef76f08f65f)
![Screenshot from 2023-08-24 20-10-15](https://github.com/Taitava/obsidian-shellcommands/assets/1771485/99faa088-64f1-459b-aba8-27029d51fa28)
![Screenshot from 2023-08-24 20-29-47](https://github.com/Taitava/obsidian-shellcommands/assets/1771485/4f0c0424-427d-4b79-bb27-33b0ad517756)

**Reply by Taitava** [reply]
Sorry, what actually is `cntrl >_`? 🙂 

How did you install Obsidian on your Linux system? I just remembered that someone had [a problem where `python` version was too old, because they had installed Obsidian from Flathub](https://github.com/Taitava/obsidian-shellcommands/discussions/225):
> I'm not sure what is the cause of this problem, may be it's because of Flatpak apps run in isolated env inside the OS, the fact that it runs python3.9 is hinting at this.

I know your problem is different: you are missing some binaries in `/usr/bin`, but perhaps the cause could be similar: maybe you are running Obsidian in some kind of isolated environment? I have [installed Obsidian on my Linux machine as an AppImage](https://help.obsidian.md/Getting+started/Download+and+install+Obsidian#Install+Obsidian+using+AppImage), and `gedit` command works correctly for me.

Please let me know if this was of any help, or if you still encounter problems. 🙂

**Reply by Pumpapa** [reply, marked as answer]
I'm sorry to be unclear. By `cntrl >_` I meant that in `Settings / Community plugins / Shell commands` there is next to each alias a menu `>_` which says (in the tooltip) `Ctrl + click: execute and ask what to do with the output`. The screenshots show the output of that.

It's installed as a flatpak
`flatpak list => Obsidian	md.obsidian.Obsidian	1.3.7	stable	user`

I wouldn't say I'm missing binaries in `/usr/bin`. They are there, but Obsidian doesn't see them. Perhaps that's what you meant.

I uninstalled Obsidian and re-installed it from .deb

Lo and behold: now it works as expected -- all commands in `/usr/bin` are visible in Shell Commands. It's annoying for you that the restricted view from a flatpak shows itself from your plugin, which is otherwise so useful and complete. I would suggest placing the caveat (which is now halfway down the README) slightly more prominently. But that's up to you, of course.

In any event, my problem was solved, and I thank you for taking the time to look into it.

**Reply by Taitava** [reply]
Nice that you got it working! 🙂

**Reply by Taitava** [reply]
> I wouldn't say I'm missing binaries in `/usr/bin`. They are there, but Obsidian doesn't see them. Perhaps that's what you meant.

Yes, that's what I meant, I just picked the wrong words. 🙂 

> I would suggest placing the caveat (which is now halfway down the README) slightly more prominently. But that's up to you, of course.

I raised it up a bit in README.md. I also [asked for help in Discord so that the plugin could detect if Obsidian was installed using Flatpak](https://discord.com/channels/686053708261228577/840286264964022302/1145751445812027553). The plugin could then show a tiny notice in the settings modal, warning users that the shell commands might not work as expected when Flatpak is used. Will see if such a feature could be developed.

**Reply by Pumpapa** [reply]
Well, thanks again for your ongoing support :+1:

**Reply by Taitava** [reply]
SC `0.21.0` now displays a warning in the settings panel, if Obsidian is installed using Flatpak.
